File: Android/app/src/main/java/jp/co/riso/android/util/Logger.java
/*
 * Copyright (c) 2014 RISO, Inc. All rights reserved.
 *
 * Logger.java
 * SmartDeviceApp
 * Created by: a-LINK Group
 */

package jp.co.riso.android.util;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.IllegalFormatException;
import java.util.Locale;

import android.content.Context;
import android.util.Log;

import jp.co.riso.smartdeviceapp.common.BaseTask;

/**
 * @class Logger
 * 
 * @brief Utility class for logging operations
 */
public class Logger {
    /// Log level to disable logs
    public static final int LOGLEVEL_NONE = Log.ASSERT + 1;
    /// Log level to enable logs for assert
    public static final int LOGLEVEL_ASSERT = Log.ASSERT;
    /// Log level to enable logs for error
    public static final int LOGLEVEL_ERROR = Log.ERROR;
    /// Log level to enable logs for warning
    public static final int LOGLEVEL_WARN = Log.WARN;
    /// Log level to enable logs for info
    public static final int LOGLEVEL_INFO = Log.INFO;
    /// Log level to enable logs for debug
    public static final int LOGLEVEL_DEBUG = Log.DEBUG;
    /// Log level to enable logs for verbose
    public static final int LOGLEVEL_VERBOSE= Log.VERBOSE;
    
    /// Log directory
    public static final String CONST_LOGS_DIR = "logs";
    /// Log filename
    public static final String CONST_TXT_FILENAME = "log.txt";
    
    private static int sLogLevel = LOGLEVEL_NONE;
    private static boolean sPerfLogs = false;
    private static boolean sPerfLogsToFile = false;
    
    private static String sStringFolder = "";
    private static HashMap<String, Long> sTimeLog = new HashMap<String, Long>();
    private static HashMap<String, Float> sMemoryLog = new HashMap<String, Float>();
    
    /**
     * @brief Initialize log.
     * 
     * @param logLevel Log level
     * @param perfLogs Enable log to logger (logcat)
     * @param perfLogsToFile Enable log to file 
     */
    public static void initialize(int logLevel, boolean perfLogs, boolean perfLogsToFile) {
        sLogLevel = logLevel;
        sPerfLogs = perfLogs;
        sPerfLogsToFile = perfLogsToFile;
        sTimeLog.clear();
        sMemoryLog.clear();
        
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMdd-HHmmssSS", Locale.getDefault());
        sStringFolder = dateFormat.format(new Date());
    }
    
    /**
     * @brief Logs a VERBOSE message. <br>
     * 
     * Logs a formatted string, using the supplied format and arguments.
     * 
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    public static void logVerbose(Class<?> cls, String format, Object...args) {
        if (cls == null || sLogLevel > LOGLEVEL_VERBOSE) {
            return;
        }

        String msg = Logger.formatMessage(format, args);
        Log.v(cls.getSimpleName(), msg);
    }

    /**
     * @brief Logs a DEBUG message. <br>
     * 
     * Logs a formatted string, using the supplied format and arguments.
     * 
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    public static void logDebug(Class<?> cls, String format, Object...args) {
        if (cls == null || sLogLevel > LOGLEVEL_DEBUG) {
            return;
        }

        String msg = Logger.formatMessage(format, args);
        Log.d(cls.getSimpleName(), (msg == null) ? "" : msg);
    }

    /**
     * @brief Logs a INFO message. <br>
     * 
     * Logs a formatted string, using the supplied format and arguments.
     * 
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    public static void logInfo(Class<?> cls, String format, Object...args) {
        if (cls == null || sLogLevel > LOGLEVEL_INFO) {
            return;
        }

        String msg = Logger.formatMessage(format, args);
        Log.i(cls.getSimpleName(), msg);
    }

    /**
     * @brief Logs a WARN message. <br>
     * 
     * Logs a formatted string, using the supplied format and arguments.
     * 
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    public static void logWarn(Class<?> cls, String format, Object...args) {
        if (cls == null || sLogLevel > LOGLEVEL_WARN) {
            return;
        }

        String msg = Logger.formatMessage(format, args);
        Log.w(cls.getSimpleName(), msg);
    }

    /**
     * @brief Logs a ERROR message. <br>
     * 
     * Logs a formatted string, using the supplied format and arguments.
     * 
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    public static void logError(Class<?> cls, String format, Object...args) {
        if (cls == null || sLogLevel > LOGLEVEL_ERROR) {
            return;
        }

        String msg = Logger.formatMessage(format, args);
        Log.e(cls.getSimpleName(), msg);
    }

    /**
     * @brief Logs a WTF message. <br>
     * 
     * Logs a formatted string, using the supplied format and arguments.
     * 
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    public static void logAssert(Class<?> cls, String format, Object...args) {
        if (cls == null || sLogLevel > LOGLEVEL_ASSERT) {
            return;
        }

        String msg = Logger.formatMessage(format, args);
        Log.wtf(cls.getSimpleName(), msg);
    }
    
    /**
     * @brief Starts a Log Time sequence.
     * 
     * @param context Valid context
     * @param cls Calling class, will be used as identifier
     * @param processName Process Name, will be used as identifier
     */
    public static void logStartTime(Context context, Class<?> cls, String processName) {
        if (!sPerfLogs || context == null || cls == null || processName == null || processName.isEmpty()) {
            return;
        }

        String tag = String.format("[%s][%s]", cls.getSimpleName(), processName);
        
        if (sTimeLog.containsKey(tag)) {
            Logger.logWarn(Logger.class, "Unterminated key, removing");
            sTimeLog.remove(tag);
            sMemoryLog.remove(tag);
        }

        long currentTime = System.currentTimeMillis();
        sTimeLog.put(tag, currentTime);

        float availableMemory = MemoryUtils.getAvailableMemory(context);
        sMemoryLog.put(tag, availableMemory);
        
        String msg = String.format(Locale.getDefault(), "%s: Started(%d) Memory(%.2f)", tag, currentTime, availableMemory);
        Logger.logInfo(Logger.class, msg);
        
        if (sPerfLogsToFile) {
            writeToFile(context, msg);
        }
    }
    
    /**
     * @brief Stops a Log Time sequence.
     * 
     * @param context Valid context
     * @param cls Calling class, will be used as identifier
     * @param processName Process Name, will be used as identifier
     */
    public static void logStopTime(Context context, Class<?> cls, String processName) {
        if (!sPerfLogs || context == null || cls == null) {
            return;
        }

        String tag = String.format("[%s][%s]", cls.getSimpleName(), processName);

        if (!sTimeLog.containsKey(tag)) {
            Logger.logError(Logger.class, "Key not started");
            return;
        }
        long startTime = sTimeLog.get(tag);
        long stopTime = System.currentTimeMillis();
        
        float startMemory = sMemoryLog.get(tag);
        float endMemory = MemoryUtils.getAvailableMemory(context);
        
        sTimeLog.remove(tag);
        sMemoryLog.remove(tag);

        String msg = String.format(Locale.getDefault(), "%s: Stopped(%d) Memory(%.2f)",
                tag, stopTime, endMemory);
        String summary = String.format(Locale.getDefault(), "%s: Duration(%.2f) Memory change(%.2f)",
                tag, (stopTime - startTime) / 1000.0f, endMemory - startMemory);
        
        Logger.logInfo(Logger.class, msg);
        Logger.logInfo(Logger.class, summary);
        
        if (sPerfLogsToFile) {
            writeToFile(context, msg);
            writeToFile(context, summary);
        }
    }
    
    /**
     * @brief Gets the entire log string from file.
     * 
     * @param context Valid context
     * 
     * @return String containing the log
     */
    public static String getLogString(Context context) {
        if (context == null) {
            return null;
        }
        
        File path = new File(getFolderString(context), CONST_TXT_FILENAME);
        
        try {
            BufferedReader in = new BufferedReader(new FileReader(path));
            StringBuffer buf = new StringBuffer();
            String str;
            while ((str = in.readLine()) != null) {
                buf.append(str);
            }
            in.close();
            return buf.toString();
        } catch (IOException e) {
            Log.e(Logger.class.getSimpleName(), "Error reading logs");
        }
        
        return null;
    }
    
    /**
     * @brief Writes the message to the log text file.
     * 
     * @param context Valid context
     * @param msg Message to be written
     */
    protected static void writeToFile(Context context, String msg) {
        if (context == null) {
            return;
        }
        
        File path = new File(getFolderString(context), CONST_TXT_FILENAME);

        Log.e(Logger.class.getSimpleName(), path.toString());
        try {
            PrintWriter pw = new PrintWriter(new FileOutputStream(path, true));
            pw.append((msg == null) ? "" : msg);
            pw.append('\n');
            pw.close();
        } catch (FileNotFoundException e) { // openFileOutput
            Log.e(Logger.class.getSimpleName(), "Error writing to file");
        }
    }
    
    /**
     * @brief Get the current logger's folder string.
     * 
     * @param context Valid context
     * @return File containing the folder path.
     */
    protected static File getFolderString(Context context) {
        if (context == null) {
            return null;
        }
        
        return context.getExternalFilesDir(CONST_LOGS_DIR + "/" + sStringFolder);
    }

    /**
     * @brief Creates the log message, and throws unnecessary exceptions.
     * 
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    private static String formatMessage(String format, Object...args) {
        try {
            String msg = (format == null) ? "" : String.format(Locale.getDefault(), format, args);
            return msg;
        } catch(IllegalFormatException e) {
            Log.e(Logger.class.getSimpleName(), "IllegalFormatException, logging format directly");
            return format;
        }
    }

    /**
     * @brief Runs a background task which deletes the application logs.
     * 
     * @param context Application context
     */
    public static void runDeleteTask(Context context) {
        if (context == null) {
            return;
        }
        
        Logger logger = new Logger();
        DeleteTask task = (new DeleteTask());
        
        task.execute(context);
    }
    
    /**
     * @class DeleteTask
     * 
     * @brief Async Task for deleting a directory
     */
    public static class DeleteTask extends BaseTask<Context, Void> {
        int count = 0;
        double time = 0;
        
        /**
         * @brief Delete specified directory.
         * 
         * Based on: http://stackoverflow.com/questions/5701586/delete-a-folder-on-sd-card 
         * 
         * @param path File object of the Directory path to delete
         */
        public boolean deleteDirectory(File path) {
            if (path.exists()) {
                File[] files = path.listFiles();
                if (files == null) {
                    return true;
                }
                for (int i = 0; i < files.length; i++) {
                    if (files[i].isDirectory()) {
                        deleteDirectory(files[i]);
                    }
                    files[i].delete();
                    count++;
                }
            }
            return true;
        }
        
        /**
         * @brief Deletes the sub folders inside the AppLogDirectory.
         * 
         * Based on: http://stackoverflow.com/questions/5701586/delete-a-folder-on-sd-card
         * 
         * @param path File object of the Directory path to delete
         */
        public boolean deleteAppLogDirectory(File path) {
            if (path.exists()) {
                File[] files = path.listFiles();
                if (files == null) {
                    return true;
                }
                for (int i = 0; i < files.length; i++) {
                    if (files[i].isDirectory()) {
                        //deletes the whole sub-directory
                        if (deleteDirectory(files[i])) {
                            files[i].delete();
                        }
                    }
                }
            }
            return true;
        }
        
        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            
            time = System.currentTimeMillis();
        }
        
        @Override
        protected Void doInBackground(Context... params) {
            deleteAppLogDirectory(params[0].getExternalFilesDir(CONST_LOGS_DIR));
            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            super.onPostExecute(result);
            
            Log.d(DeleteTask.class.getSimpleName(), "Duration: " + (System.currentTimeMillis() - time));
            Log.d(DeleteTask.class.getSimpleName(), "Count: " + count);
        }
    }
    
    static {
        initialize(LOGLEVEL_NONE, false, false);
    }
}

File: Android/app/src/main/java/jp/co/riso/android/util/Logger.kt
/*
 * Copyright (c) 2014 RISO, Inc. All rights reserved.
 *
 * Logger.java
 * SmartDeviceApp
 * Created by: a-LINK Group
 */
package jp.co.riso.android.util

import android.content.Context
import android.util.Log
import jp.co.riso.android.util.MemoryUtils
import jp.co.riso.android.util.Logger.DeleteTask
import jp.co.riso.smartdeviceapp.common.BaseTask
import java.io.*
import java.text.SimpleDateFormat
import java.util.*

/**
 * @class Logger
 *
 * @brief Utility class for logging operations
 */
object Logger {
    /// Log level to disable logs
    const val LOGLEVEL_NONE = Log.ASSERT + 1

    /// Log level to enable logs for assert
    const val LOGLEVEL_ASSERT = Log.ASSERT

    /// Log level to enable logs for error
    const val LOGLEVEL_ERROR = Log.ERROR

    /// Log level to enable logs for warning
    const val LOGLEVEL_WARN = Log.WARN

    /// Log level to enable logs for info
    const val LOGLEVEL_INFO = Log.INFO

    /// Log level to enable logs for debug
    const val LOGLEVEL_DEBUG = Log.DEBUG

    /// Log level to enable logs for verbose
    const val LOGLEVEL_VERBOSE = Log.VERBOSE

    /// Log directory
    const val CONST_LOGS_DIR = "logs"

    /// Log filename
    const val CONST_TXT_FILENAME = "log.txt"
    private var sLogLevel = LOGLEVEL_NONE
    private var sPerfLogs = false
    private var sPerfLogsToFile = false
    private var sStringFolder = ""
    private val sTimeLog = HashMap<String, Long>()
    private val sMemoryLog = HashMap<String, Float>()

    /**
     * @brief Initialize log.
     *
     * @param logLevel Log level
     * @param perfLogs Enable log to logger (logcat)
     * @param perfLogsToFile Enable log to file
     */
    @JvmStatic
    fun initialize(logLevel: Int, perfLogs: Boolean, perfLogsToFile: Boolean) {
        sLogLevel = logLevel
        sPerfLogs = perfLogs
        sPerfLogsToFile = perfLogsToFile
        sTimeLog.clear()
        sMemoryLog.clear()
        val dateFormat = SimpleDateFormat("yyyyMMdd-HHmmssSS", Locale.getDefault())
        sStringFolder = dateFormat.format(Date())
    }

    /**
     * @brief Logs a VERBOSE message. <br></br>
     *
     * Logs a formatted string, using the supplied format and arguments.
     *
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    @JvmStatic
    fun logVerbose(cls: Class<*>?, format: String?, vararg args: Any) {
        if (cls == null || sLogLevel > LOGLEVEL_VERBOSE) {
            return
        }
        val msg = formatMessage(format, *args)
        Log.v(cls.simpleName, msg!!)
    }

    /**
     * @brief Logs a DEBUG message. <br></br>
     *
     * Logs a formatted string, using the supplied format and arguments.
     *
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    @JvmStatic
    fun logDebug(cls: Class<*>?, format: String?, vararg args: Any) {
        if (cls == null || sLogLevel > LOGLEVEL_DEBUG) {
            return
        }
        val msg = formatMessage(format, *args)
        Log.d(cls.simpleName, msg ?: "")
    }

    /**
     * @brief Logs a INFO message. <br></br>
     *
     * Logs a formatted string, using the supplied format and arguments.
     *
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    @JvmStatic
    fun logInfo(cls: Class<*>?, format: String?, vararg args: Any) {
        if (cls == null || sLogLevel > LOGLEVEL_INFO) {
            return
        }
        val msg = formatMessage(format, *args)
        Log.i(cls.simpleName, msg!!)
    }

    /**
     * @brief Logs a WARN message. <br></br>
     *
     * Logs a formatted string, using the supplied format and arguments.
     *
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    @JvmStatic
    fun logWarn(cls: Class<*>?, format: String?, vararg args: Any) {
        if (cls == null || sLogLevel > LOGLEVEL_WARN) {
            return
        }
        val msg = formatMessage(format, *args)
        Log.w(cls.simpleName, msg!!)
    }

    /**
     * @brief Logs a ERROR message. <br></br>
     *
     * Logs a formatted string, using the supplied format and arguments.
     *
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    @JvmStatic
    fun logError(cls: Class<*>?, format: String?, vararg args: Any) {
        if (cls == null || sLogLevel > LOGLEVEL_ERROR) {
            return
        }
        val msg = formatMessage(format, *args)
        Log.e(cls.simpleName, msg!!)
    }

    /**
     * @brief Logs a WTF message. <br></br>
     *
     * Logs a formatted string, using the supplied format and arguments.
     *
     * @param cls Calling class. Will be used as Log's TAG
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    @JvmStatic
    fun logAssert(cls: Class<*>?, format: String?, vararg args: Any) {
        if (cls == null || sLogLevel > LOGLEVEL_ASSERT) {
            return
        }
        val msg = formatMessage(format, *args)
        Log.wtf(cls.simpleName, msg)
    }

    /**
     * @brief Starts a Log Time sequence.
     *
     * @param context Valid context
     * @param cls Calling class, will be used as identifier
     * @param processName Process Name, will be used as identifier
     */
    @JvmStatic
    fun logStartTime(context: Context?, cls: Class<*>?, processName: String?) {
        if (!sPerfLogs || context == null || cls == null || processName == null || processName.isEmpty()) {
            return
        }
        val tag = String.format("[%s][%s]", cls.simpleName, processName)
        if (sTimeLog.containsKey(tag)) {
            logWarn(Logger::class.java, "Unterminated key, removing")
            sTimeLog.remove(tag)
            sMemoryLog.remove(tag)
        }
        val currentTime = System.currentTimeMillis()
        sTimeLog[tag] = currentTime
        val availableMemory = MemoryUtils.getAvailableMemory(context)
        sMemoryLog[tag] = availableMemory
        val msg = String.format(
            Locale.getDefault(),
            "%s: Started(%d) Memory(%.2f)",
            tag,
            currentTime,
            availableMemory
        )
        logInfo(Logger::class.java, msg)
        if (sPerfLogsToFile) {
            writeToFile(context, msg)
        }
    }

    /**
     * @brief Stops a Log Time sequence.
     *
     * @param context Valid context
     * @param cls Calling class, will be used as identifier
     * @param processName Process Name, will be used as identifier
     */
    @JvmStatic
    fun logStopTime(context: Context?, cls: Class<*>?, processName: String?) {
        if (!sPerfLogs || context == null || cls == null) {
            return
        }
        val tag = String.format("[%s][%s]", cls.simpleName, processName)
        if (!sTimeLog.containsKey(tag)) {
            logError(Logger::class.java, "Key not started")
            return
        }
        val startTime = sTimeLog[tag]!!
        val stopTime = System.currentTimeMillis()
        val startMemory = sMemoryLog[tag]!!
        val endMemory = MemoryUtils.getAvailableMemory(context)
        sTimeLog.remove(tag)
        sMemoryLog.remove(tag)
        val msg = String.format(
            Locale.getDefault(), "%s: Stopped(%d) Memory(%.2f)",
            tag, stopTime, endMemory
        )
        val summary = String.format(
            Locale.getDefault(), "%s: Duration(%.2f) Memory change(%.2f)",
            tag, (stopTime - startTime) / 1000.0f, endMemory - startMemory
        )
        logInfo(Logger::class.java, msg)
        logInfo(Logger::class.java, summary)
        if (sPerfLogsToFile) {
            writeToFile(context, msg)
            writeToFile(context, summary)
        }
    }

    /**
     * @brief Gets the entire log string from file.
     *
     * @param context Valid context
     *
     * @return String containing the log
     */
    @JvmStatic
    fun getLogString(context: Context?): String? {
        if (context == null) {
            return null
        }
        val path = File(getFolderString(context), CONST_TXT_FILENAME)
        try {
            val `in` = BufferedReader(FileReader(path))
            val buf = StringBuffer()
            var str: String?
            while (`in`.readLine().also { str = it } != null) {
                buf.append(str)
            }
            `in`.close()
            return buf.toString()
        } catch (e: IOException) {
            Log.e(Logger::class.java.simpleName, "Error reading logs")
        }
        return null
    }

    /**
     * @brief Writes the message to the log text file.
     *
     * @param context Valid context
     * @param msg Message to be written
     */
    @JvmStatic
    fun writeToFile(context: Context?, msg: String?) {
        if (context == null) {
            return
        }
        val path = File(getFolderString(context), CONST_TXT_FILENAME)
        Log.e(Logger::class.java.simpleName, path.toString())
        try {
            val pw = PrintWriter(FileOutputStream(path, true))
            pw.append(msg ?: "")
            pw.append('\n')
            pw.close()
        } catch (e: FileNotFoundException) { // openFileOutput
            Log.e(Logger::class.java.simpleName, "Error writing to file")
        }
    }

    /**
     * @brief Get the current logger's folder string.
     *
     * @param context Valid context
     * @return File containing the folder path.
     */
    internal fun getFolderString(context: Context?): File? {
        return context?.getExternalFilesDir(CONST_LOGS_DIR + "/" + sStringFolder)
    }

    /**
     * @brief Creates the log message, and throws unnecessary exceptions.
     *
     * @param format The format string
     * @param args The list of arguments passed to the formatter. If there are more arguments than required by format, additional arguments are ignored.
     */
    private fun formatMessage(format: String?, vararg args: Any): String? {
        return try {
            if (format == null) "" else String.format(Locale.getDefault(), format, *args)
        } catch (e: IllegalFormatException) {
            Log.e(Logger::class.java.simpleName, "IllegalFormatException, logging format directly")
            format
        }
    }

    /**
     * @brief Runs a background task which deletes the application logs.
     *
     * @param context Application context
     */
    @JvmStatic
    fun runDeleteTask(context: Context?) {
        if (context == null) {
            return
        }
        val task = DeleteTask()
        task.execute(context)
    }

    /**
     * @class DeleteTask
     *
     * @brief Async Task for deleting a directory
     */
    class DeleteTask : BaseTask<Context?, Void?>() {
        var count = 0
        var time = 0.0

        /**
         * @brief Delete specified directory.
         *
         * Based on: http://stackoverflow.com/questions/5701586/delete-a-folder-on-sd-card
         *
         * @param path File object of the Directory path to delete
         */
        fun deleteDirectory(path: File): Boolean {
            if (path.exists()) {
                val files = path.listFiles() ?: return true
                for (file in files) {
                    if (file.isDirectory) {
                        deleteDirectory(file)
                    }
                    file.delete()
                    count++
                }
            }
            return true
        }

        /**
         * @brief Deletes the sub folders inside the AppLogDirectory.
         *
         * Based on: http://stackoverflow.com/questions/5701586/delete-a-folder-on-sd-card
         *
         * @param path File object of the Directory path to delete
         */
        fun deleteAppLogDirectory(path: File?): Boolean {
            if (path!!.exists()) {
                val files = path.listFiles() ?: return true
                for (file in files) {
                    if (file.isDirectory) {
                        //deletes the whole sub-directory
                        if (deleteDirectory(file)) {
                            file.delete()
                        }
                    }
                }
            }
            return true
        }

        override fun onPreExecute() {
            super.onPreExecute()
            time = System.currentTimeMillis().toDouble()
        }

        protected override fun doInBackground(vararg params: Context?): Void? {
            deleteAppLogDirectory(params[0]?.getExternalFilesDir(CONST_LOGS_DIR))
            return null
        }

        protected override fun onPostExecute(result: Void?) {
            super.onPostExecute(result)
            Log.d(
                DeleteTask::class.java.simpleName,
                "Duration: " + (System.currentTimeMillis() - time)
            )
            Log.d(DeleteTask::class.java.simpleName, "Count: $count")
        }

    }

    init {
        initialize(LOGLEVEL_NONE, false, false)
    }
}